org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/preferences/LifecycleMappingsViewer.java

```diff
@@ -444,52 +475,56 @@ String getSourcelabel(MojoExecutionKey execution, List<IPluginExecutionMetadata>
     LinkedHashSet<String> sources = new LinkedHashSet<>();
     if(mappings != null && !mappings.isEmpty()) {
       for(IPluginExecutionMetadata mapping : mappings) {
-        LifecycleMappingMetadataSource metadata = ((PluginExecutionMetadata) mapping).getSource();
-        if(metadata != null) {
-          Object source = metadata.getSource();
-          if(source instanceof String s) {
-            sources.add(s);
-          } else if(source instanceof Artifact artifact) {
-            sources.add(getSourceLabel(artifact, detailed));
-          } else if(source instanceof MavenProject mavenProject) {
-            sources.add(getSourceLabel(mavenProject, detailed));
-          } else if(source instanceof Bundle bundle) {
-            sources.add(getSourceLabel(bundle, detailed));
+        if(mapping instanceof PluginExecutionMetadata metadata) {
+          LifecycleMappingMetadataSource metadataSource = metadata.getSource();
+          if(metadataSource == null) {
+            sources.add(UNKNOWN);
           } else {
-            sources.add("unknown"); //$NON-NLS-1$
+            Object source = metadataSource.getSource();
+            if(source instanceof String s) {
+              sources.add(s);
+            } else if(source instanceof Artifact artifact) {
+              sources.add(getSourceLabel(artifact, detailed));
+            } else if(source instanceof MavenProject mavenProject) {
+              sources.add(getSourceLabel(mavenProject, detailed));
+            } else if(source instanceof Bundle bundle) {
+              sources.add(getSourceLabel(bundle, detailed));
+            } else {
+              sources.add(UNKNOWN); //$NON-NLS-1$
+            }
           }
         } else if(mapping instanceof DefaultPluginExecutionMetadata) {
```

Comment:
This path must be outside the `if(mapping instanceof PluginExecutionMetadata metadata) {` block.

Reply:
It is already.
```java
if (mapping instanceof PluginExecutionMetadata) {
    if (source == null) {
        // unknown
    } else {
        // source label or unknown
    }
} else if (mapping instanceof DefaultPluginExecutionMetadata) {
    // default
} else {
    // unknown
}

Reply:
Accidentally replied on the wrong account, but this is how it is.

```java
if (mapping instanceof PluginExecutionMetadata) {
    if (source == null) {
        // unknown
    } else {
        // source label or unknown
    }
} else if (mapping instanceof DefaultPluginExecutionMetadata) {
    // default
} else {
    // unknown
}
```